### PR TITLE
Edit  mangas' Categories in Library using TriState list

### LINKED ISSUE #961: Handling of multiple categories, might remove categories
When you select multiple series and then change categories, the categories which not all series have will be removed. It seems the cagegory-dialog only gets the list of categories which all series have shared. On save it writes those categories back to all selected series, overwriting the existing list, instead of just adding or removing changed categories.

Example:

    Manga_1 with [Categorie_A]
    Manga_2 with [Categorie_A, Categorie_B]

becomes after selection and saving:

    Manga_1 with [Categorie_A]
    Manga_2 with [Categorie_A]

Version: 0.6.1
Android 6.0

### PR DESCRIPTION
**Pull request by quangkieu**
Fix #961
Add updateMangasToCategories to build build correct Categories list for
  each manga using Common and Mix list
Update QuadState Multi-Choice to be either Action or Display List
  Display list would have different state sequence from Action
  Uncheck-> Indeterminate (only if initial so)-> Check

Signed-off-by: Quang Kieu quangkieu1993@gmail.com

### CODE REVIEW

**Comment by CarlosEsco** [comment]
Can you provide screens of the change

**Comment by quangkieu** [comment]
<details>
<summary>Image from my phone</summary>

Screenshot_20210618-231824_Tachiyomi


</details>

<details>
<summary>Video from Bluestack</summary>


</details>
The video show case the feature of tri-state box and remember its mix status, and how the code behind handle mix, common categories so that we do not loss the info anymore

**Comment by quangkieu** [comment]
I am not at all develop in android/kotlin before. I just copy the code that @arkon mention last time and mix them up to fit.
I would happy if someone could look if any of these follow the prefer practice of android philosophy

**Comment by arkon** [comment]
Before I really dive into it, it'd probably actually be best to just modify the existing dialog rather than creating a new sheet. Were you concerned by differing behaviour between using it in the library and in browse or something?

**Comment by quangkieu** [comment]
So mostly that I do not know how to modify dialog to add tristate so I just copy paste the current implemented classes that using it and expand from there. So I base my design from the library display setting sheet

**Comment by arkon** [comment]
The category-related settings dialogs (global updates and automatic downloads) both implement tristate (actually quadstate) checkboxes, so you could look at that.

e.g. https://github.com/tachiyomiorg/tachiyomi/blob/c3fb5af3fcbeee4f3579435b832907d8165cb42e/app/src/main/java/eu/kanade/tachiyomi/ui/setting/SettingsLibraryController.kt#L359

**Comment by quangkieu** [comment]
I could look into it but I wonder if you prefer same UI/UI in Library between catalog and display setting or you would like them the same as manga/browse section dialog?
As I could see we slowly move things to sheets new design
Also I would likely need to update listener signature so I could send both common and mix list to build final catalogs per manga

**Comment by quangkieu** [comment]
From digging more into the class and try to implement it, the listItemsQuadStateMultiChoice which using QuadStateMultiChoiceDialogAdapterhas its state sequence hardcoded from Uncheck, Check, Invert. So look like I have to build new dialog class to define the state sequence.
From the outlook it is not bring much to the table and the TabbedBottomSheetDialog which use ExtendedNavigationView has more hook and interface for easy define new classes

**Comment by quangkieu** [comment]
I wonder if there anything I need to change?

**Comment by arkon** [comment]
I don't want to go down the TabbedBottomSheetDialog/ExtendedNavigationView route right now since:
- It kind of looks bad, since it's not really a tabbed experience
- We'd have to transition more things to sheets to be consistent, which I want to avoid right now
- If we _do_ use sheets, I'd rather it be more customized, i.e. similar to what's currently in TachiyomiJ2K

It shouldn't be hard to modify `listItemsQuadStateMultiChoice`/`QuadStateMultiChoiceDialogAdapter` to allow changing the behaviour of what state it goes to (i.e. selected to unselected instead of the current selected to inversed path). May want to wait for #5423 to be completed first though.

**Comment by vengiare** [comment]
I've discussed this in #3382 before and was just wondering (can't test the PR myself) if it handles multiple indeterminate states just fine. Say you select a manga from two different categories.

An example would be

Category A
– Manga 1
Category B
– Manga 2
Category C
– <blank>

And then turn it to

Category A (indeterminate)
– Manga 1
Category B (indeterminate to checked)
– +Manga 1
– Manga 2
Category C (unchecked to checked)
– +Manga 1
– +Manga 2

**Comment by quangkieu** [comment]
New version using QuadStateMultiChoiceDialogAdapter
QuadStateMultiChoiceDialogAdapter now have two types of List: Display, Action
Library categories display list has same behavior as previous version. State sequence: Uncheck-> Indeterminate (only if initial so)-> Check
image

File: app/src/main/java/eu/kanade/tachiyomi/ui/library/LibraryPresenter.kt [hunk 1]
```diff
@@ -533,4 +545,22 @@ class LibraryPresenter(
...
+    /**
+     * Bulk update categories of mangas using old and new common categories.
+     *
+     * @param mangas the list of manga to move.
+     * @param newCommon the selected categories.
+     * @param newMix the list of common categories pre-update
+     */
+    fun updateMangasToCategories(mangas: List<Manga>, newCommon: List<Category>, newMix: List<Category>) {
+        val mc = mutableListOf<MangaCategory>()
+        for (manga in mangas) {
```
Comment (arkon) [inline app/src/main/java/eu/kanade/tachiyomi/ui/library/LibraryPresenter.kt:558]: I think you could do something along the lines of

```
mangas.map {
  val categories = ...
  MangaCategory.create(it, categories) 
}
```

File: app/src/main/java/eu/kanade/tachiyomi/ui/manga/MangaController.kt [hunk 2]
```diff
@@ -626,15 +631,13 @@ class MangaController :
         val categories = presenter.getCategories()
 
         val ids = presenter.getMangaCategoryIds(manga)
-        val preselected = ids.mapNotNull { id ->
-            categories.indexOfFirst { it.id == id }.takeIf { it != -1 }
-        }.toTypedArray()
+        val preselected = categories.map { if (it.id in ids) 2 else 0 }.toTypedArray()
```
Comment (arkon) [inline app/src/main/java/eu/kanade/tachiyomi/ui/manga/MangaController.kt:634]: Same here re: enum usage

**Review by arkon — COMMENTED** [review]
Looks fine otherwise.

File: app/src/main/java/eu/kanade/tachiyomi/widget/materialdialogs/QuadStateMultiChoiceDialogAdapter.kt [hunk 3]
```diff
@@ -56,6 +67,21 @@ internal class QuadStateMultiChoiceDialogAdapter(
         listener(currentSelection)
     }
 
+    internal fun itemDisplayClicked(index: Int) {
+        val newSelection = this.currentSelection.toMutableList()
+        newSelection[index] = when (currentSelection[index]) {
+            QuadStateTextView.State.UNCHECKED.ordinal -> when (initialSelected[index]) {
+                QuadStateTextView.State.INDETERMINATE.ordinal -> QuadStateTextView.State.INDETERMINATE.ordinal
```
[the remark this answers is not in the document]
Comment (quangkieu) [reply app/src/main/java/eu/kanade/tachiyomi/widget/materialdialogs/QuadStateMultiChoiceDialogAdapter.kt:71]: So is this resolved then

**Comment by arkon** [comment]
Thanks!

**Pull request merged**